Add a MovingFunction pipeline aggregation, deprecate MovingAvg agg

polyfractal opened the pull request:
This pipeline aggregation gives the user the ability to script functions that "move" across a window
of data, instead of single data points. Through custom script contexts, we also expose a number of convenience methods:
 - `windowMax()`
 - `windowMin()`
 - `windowSum()`
 - `simpleMovAvg()`
 - `linearMovAvg()`
 - `ewmaMovAvg()`
 - `holtMovAvg()`
 - `holtWintersMovAvg()`

Since we can easily expose moving average functionality through custom script contexts, I would like to deprecate the `moving_avg` agg at the same time if there isn't objection.  Thoughts?

At the moment, this agg doesn't support "predictions" like the old agg... but I think that is a good thing.  The "prediction" functionality is just super basic extrapolation, is fairly confusing for users and of questionable value.  It also means we can eventually remove the simulated annealing optimizer, since that was mainly there to help minimize the parameters for holt-winters prediction.

I gutted the math portions of the moving avg models and consolidated them in this PR.  There's still a lot of boilerplate cruft from the old models left, but that's just the way it is.  In the process I de-generified the models to make it work with MovFn, and generics weren't needed anyway.

This is still a WIP, needs docs, etc.  @rjernst would you mind taking a look at the scripting context stuff to make sure it looks ok?  I'll track down a search/aggs team member for the rest of the review when it's ready.

This is a reboot of #25137

elasticmachine commented:
Pinging @elastic/es-search-aggs

polyfractal commented on server/src/main/java/org/elasticsearch/search/aggregations/pipeline/BucketHelpers.java:
```diff
@@ -79,7 +79,7 @@ public static GapPolicy parse(String text, XContentLocation tokenLocation) {
                     validNames.add(policy.getName());
                 }
-                throw new ParsingException(tokenLocation, "Invalid gap policy: [" + text + "], accepted values: " + validNames);
+                throw new IllegalArgumentException("Invalid gap policy: [" + text + "], accepted values: " + validNames);
```
I thought it was incompatible with the new parser style  (I didnt think you could get token info with the new way of writing parsers), but while writing this comment I noticed how BucketSort did it:

```
         PARSER.declareField(BucketSortPipelineAggregationBuilder::gapPolicy, p -> {
             if (p.currentToken() == XContentParser.Token.VALUE_STRING) {
                return GapPolicy.parse(p.text().toLowerCase(Locale.ROOT), p.getTokenLocation());
             }
             throw new IllegalArgumentException("Unsupported token [" + p.currentToken() + "]");
         }
```

which I somehow missed before.  I'll put it back to the way it was and using the above snippet :)

polyfractal commented on server/src/main/java/org/elasticsearch/search/aggregations/pipeline/movfn/MovingFunctions.java:
```diff
@@ -0,0 +1,275 @@
+     */
+    public static double linearMovAvg(Collection<Double> values) {
+        double avg = 0;
+        long totalWeight = 1;
```
Hmm, good point.  OTOH, that's how the linear avg works today so it'd be "breaking" the current behavior (these functions were basically lifted directly from existing moving model code)

I think we should probably change it anyway, but wanted to raise that point.

polyfractal commented on server/src/main/java/org/elasticsearch/search/aggregations/pipeline/movfn/MovingFunctions.java:
```diff
@@ -0,0 +1,275 @@
+                avg = (v * alpha) + (avg * (1 - alpha));
+            }
+        }
+        return avg;
```
Ditto above

colings86 commented on docs/reference/aggregations/pipeline/movfn-aggregation.asciidoc:
```diff
@@ -0,0 +1,575 @@
+etc.
+
+This is conceptually very similar to the <<search-aggregations-pipeline-movavg-aggregation, Moving Average>> pipeline aggregation, except
+it provides more functionality and better flexibility.
```
`more functionality and better flexibility` -> I think we can just say `more functionality` here rather than repeating?

colings86 commented on docs/reference/aggregations/pipeline/movfn-aggregation.asciidoc:
```diff
@@ -0,0 +1,575 @@
+|`buckets_path` |Path to the metric of interest (see <<buckets-path-syntax, `buckets_path` Syntax>> for more details |Required |
+|`window` |The size of window to "slide" across the histogram. |Optional |`5`
+|`script` |The script that should be executed on each window of data |Required |
+|`params` |User-defined extra parameters to be injected into the script |Optional |
```
This shouldn't be needed as the `script` field itself can handle params?

colings86 commented on docs/reference/aggregations/pipeline/movfn-aggregation.asciidoc:
```diff
@@ -0,0 +1,575 @@
+// TEST[setup:sales]
+
+<1> A `date_histogram` named "my_date_histo" is constructed on the "timestamp" field, with one-day intervals
+<2> A `sum` metric is used to calculate the sum of a field.  This could be any metric (sum, min, max, etc)
```
I think we should say `this could be any numeric metric as I don't think you could point a `moving_fn` at, for example, a geo_centroid agg?

colings86 commented on docs/reference/aggregations/pipeline/movfn-aggregation.asciidoc:
```diff
@@ -0,0 +1,575 @@
+<3> Finally, we specify a `moving_fn` aggregation which uses "the_sum" metric as its input.
+
+Moving averages are built by first specifying a `histogram` or `date_histogram` over a field.  You can then optionally
+add normal metrics, such as a `sum`, inside of that histogram.  Finally, the `moving_fn` is embedded inside the histogram.
```
normal -> numeric ?

colings86 commented on docs/reference/aggregations/pipeline/movfn-aggregation.asciidoc:
```diff
@@ -0,0 +1,575 @@
+- `linearMovAvg()`
+- `ewmaMovAvg()`
+- `holtMovAvg()`
+- `holtWintersMovAvg()`
```
I wonder if it would be useful to add `windowStdDev()` to this list too as I have seen some requests for moving standard deviations too. However I imagine given that it would be used together with moving average, it would need the simple, linear, ewma, holt and holtWinters forms to make sense with moving average, so maybe its too complex to add it in this PR?

colings86 commented on docs/reference/aggregations/pipeline/movfn-aggregation.asciidoc:
```diff
@@ -0,0 +1,575 @@
+===== windowMax Function
+
+This function accepts a collection of doubles and returns the maximum value in that window.  If the window is empty, or all values
+are `null`/`NaN`, `NaN` is returned as the result.
```
what happens if some of the values are `NaN`? does it ignore those `NaN` values and return the max of the others or return `NaN`?

colings86 replied:
This also applies to the other functions

polyfractal replied:
Yes, it'll find the `<function>` of the non-`NaN` values, ignoring the `NaN`s.  I reasoned that you can't find the min/max/whatever of a `NaN` because it isn't a number, so there's no way to compare it to actual numbers and they should just be ignored.

That makes the averages a bit tricky though, since they are calculating the average based on number of non-`NaN` values not "positions in the window".  It feels like if the user wants to count-but-not-use `NaN`s, they should probably write their own avg function at that point.

I'll beef up the documentation to make this point clear, since it's not right now.

Note: `+/- Inf` are treated as proper numbers for all functions, since the you can say `1 < +Inf`

colings86 commented on docs/reference/aggregations/pipeline/movfn-aggregation.asciidoc:
```diff
@@ -0,0 +1,575 @@
+===== windowMin Function
+
+This function accepts a collection of doubles and returns the minimum value in that window.  If the window is empty, or all values
+are `null`/`NaN`, `NaN` is returned as the result.
```
as above

colings86 commented on docs/reference/aggregations/pipeline/movfn-aggregation.asciidoc:
```diff
@@ -0,0 +1,575 @@
+===== windowSum Function
+
+This function accepts a collection of doubles and returns the sum of the values in that window.  If the window is empty, or all values
+are `null`/`NaN`, `0.0` is returned as the result.
```
as above

colings86 commented on server/src/main/java/org/elasticsearch/search/aggregations/pipeline/movavg/models/SimpleModel.java:
```diff
@@ -84,7 +84,11 @@ public MovAvgModel clone() {
     @Override
     public double next(Collection<Double> values) {
-        return MovingFunctions.simpleMovAvg(values);
+        double avg = 0;
```
can we call this `sum` as its not actually the `avg`? 😉

polyfractal replied:
durp.  :)

colings86 commented on server/src/main/java/org/elasticsearch/search/aggregations/pipeline/movavg/models/LinearModel.java:
```diff
@@ -86,7 +86,16 @@ public MovAvgModel clone() {
 
     @Override
     public double next(Collection<Double> values) {
```
Can we add a comment here explaining why this is duplicating functionality provided by `MovingFunctions`? I think someone could try to deduplicate the code without this

polyfractal replied:
Oof, actually, this should be pulled back out.  This was added when I thought it was a breaking change but it's not actually (since we don't allow null/NaN as a value because of `gap_policy`) in regular mov avg.

Thanks for noticing, I'll revert this to pointing at the static MovingFunctions  version.

colings86 commented on server/src/main/java/org/elasticsearch/search/aggregations/pipeline/movfn/MovingFunctions.java:
```diff
@@ -21,59 +21,83 @@
     public static double windowSum(Collection<Double> values) {
-        return values.stream().mapToDouble(Double::doubleValue).sum();
+        if (values.size() == 0) {
+            return 0.0;
```
From the docs change above it says:
> If the window is empty, or all values are `null`/`NaN`, `0.0` is returned as the result.

This seems to not be the case here?

polyfractal replied:
?  If the size is zero, it returns `0.0`.  And in the `mapToDouble()`, we only use the value if it's non-null and non-NaN (otherwise we return 0.0 for that position).

There's a test to verify this: https://github.com/elastic/elasticsearch/pull/29594/files/83778999a9aec1e0170efba0d4857100a927a770..c93d065b6ffd4e37089b4e251ad784380b75a942#diff-a35897b4e9c0a693a4515cab553a6427R73

colings86 commented on server/src/main/java/org/elasticsearch/search/aggregations/pipeline/movfn/MovFnPipelineAggregationBuilder.java:
```diff
@@ -0,0 +1,264 @@
+    private final String bucketsPathString;
+    private String format = null;
+    private GapPolicy gapPolicy = GapPolicy.SKIP;
+    private int window;
```
Since this is optional we should set the default here

polyfractal replied:
Doc error.  Window isn't optional, I'll make the fix in the documentation.

colings86 commented on server/src/main/java/org/elasticsearch/search/aggregations/pipeline/movfn/MovFnPipelineAggregationBuilder.java:
```diff
@@ -0,0 +1,264 @@
+    };
+
+
+    public MovFnPipelineAggregationBuilder(String name, String bucketsPath, Script script, int window) {
```
Since the `window` is optional we should not provide it in the constructor here but rather with a setter method.

polyfractal commented:
- Added more details in docs about handling of `NaN` and `null`, and how it affects avg counts
- Made `window` mandatory in docs
- Added a standard deviation.  I decided to sidestep the issue about what avg to use (and proliferation of a std dev to match each avg) by just making the user pass in the avg.  The example shows how to combine one of the moving avg functions with std dev to get the desired behavior.  Seemed only slightly less user-friendly, but a lot more flexible and less hassle to maintain
- Fixed up a bunch of doc testing issues

rjernst commented on server/src/main/java/org/elasticsearch/search/aggregations/pipeline/movfn/MovFnScript.java:
```diff
@@ -0,0 +1,45 @@
+ * This class provides a custom script context for the Moving Function pipeline aggregation,
+ * so that we can expose a number of pre-baked moving functions like min, max, movavg, etc
+ */
+public abstract class MovFnScript {
```
Can the name not be so cryptic? `MovingFunctionScript`?

rjernst commented on server/src/main/java/org/elasticsearch/search/aggregations/pipeline/movfn/MovingFunctions.java:
```diff
@@ -0,0 +1,360 @@
+     * If all values are missing/null/NaN, the return value will be NaN.
+     * The average is based on the count of non-null, non-NaN values.
+     */
+    public static double simpleMovAvg(Collection<Double> values) {
```
Maybe just call this `unweightedAvg`? Is the `MovAvg` suffix really necessary given this is within `MovingFunctions`? Ditto for the methods below.

rjernst commented on server/src/main/java/org/elasticsearch/search/aggregations/pipeline/movfn/MovingFunctions.java:
```diff
@@ -0,0 +1,360 @@
+     * Find the maximum value in a window of values.
+     * If all values are missing/null/NaN, the return value will be NaN
+     */
+    public static double windowMax(Collection<Double> values) {
```
What do you think of using DoubleStream as the type here? I think it could save boxing in some cases?

polyfractal commented:
More cleanup:

- Renamed `MovFnScript` to `MovingFunctionScript`
- Test fix for movAvg, needed to assertWarning in a few places
- Discussed with @colings86 about naming and we decided to overhaul all the names, not just `simpleMovAvg`.  The list is now:
  - `max()`
  - `min()`
  - `sum()`
  - `stdDev()`
  - `unweightedAvg()`
  - `linearWeightedAvg()`
  - `ewma()`
  - `holt()`
  - `holtWinters()`
  Chose to keep ewma/holt/holtWinters instead of exponentionalWeighted/doubleExponentialWeighted/tripleExponentialWeighted for brevity :)

polyfractal commented on server/src/main/java/org/elasticsearch/search/aggregations/pipeline/movfn/MovingFunctions.java:
```diff
@@ -0,0 +1,359 @@
+     * Find the maximum value in a window of values.
+     * If all values are missing/null/NaN, the return value will be NaN
+     */
+    public static double max(Collection<Double> values) {
```
I think `double[]` should be doable.  :+1: 

I didn't do that originally to avoid unboxing the internal `EvictingQueue` into an array since it wouldn't be necessary most of the time.  And I avoided specifying a `Queue` as the type since the user might want to use these functions outside of the values we provide (or in addition to).

But I didn't think about the unordered nature of `Collection`, that's definitely an issue.  Lemme poke at it tomorrow with `double[]`, I think that'll work fine.

_Sidebar_ 
Alternatively, I could replace the evicting queue with a primitive-based ring buffer.  I don't think that's any better though.  We'd still have to arrayCopy the relevant slice into the script since there are null/NaNs to deal with, and we probably shouldn't make the method signature some weird ring buffer class :)

> Note that doc['fieldname'] returns an object that implements List, but we could make this easier by adding an asArray() method to DoubleScriptDocValues?

Ah interesting.  I'm not sure it's needed... do we think users will plop a field's value into the functions instead of the values?  I was thinking users might tweak the values before a function, or write their own based on the values... but not use an entire field in place of the provided values.  Thoughts?

rjernst replied:
Forget what I said about ScriptDocValues. I was thinking this was like scripted metric agg, which is a SearchScript, but you did this correctly (no getDoc()). So no need to do anything with the doc values accessors.

polyfractal commented:
@rjernst swapped over to `double[]`.  As a side benefit, it makes things a lot easier for the user in the script (`values[0]` just seems more ergonomic than working with collections/lists) :)

polyfractal commented:
@colings86 @rjernst did either of you want another review pass on this?
